tests/test_paragraph_location.py

```
@@ -539,3 +539,62 @@ def test_stale_ref_on_long_paragraph_truncates_preview(simple_docx, tmp_path):
+
+    The defining contract: each ``(ref, loc)`` entry must be byte-identical
+    to what ``get_paragraph_location(ref)`` returns per-call. The batch path
+    only swaps the whole-DOM table rescan for a prebuilt index, so the two
+    must never disagree.
+    """
+
+    def test_returns_one_entry_per_paragraph(self, gridspan_docx):
+        with Document.open(gridspan_docx) as doc:
+            entries = doc.list_paragraph_locations()
+            assert len(entries) == len(doc.list_paragraphs())
+            assert len(entries) == 11  # P1–P11 from the fixture
```

Ruff (0.15.15)

[warning] Comment contains ambiguous `–` (EN DASH). Did you mean `-` (HYPHEN-MINUS)?

(RUF003)